benchmarks/single_node/test_single_node.py

```diff
@@ -0,0 +1,121 @@
+import numpy as np
+import ray
+import ray.autoscaler.sdk
+from ray.test_utils import Semaphore
+
+from time import sleep
+from tqdm import trange, tqdm
+
+
+MAX_RETURNS = 1000
+MAX_RAY_GET_ARGS = 10000
+MAX_QUEUED_TASKS = 1_000_000
+MAX_RAY_GET_SIZE = 100 * 2**30
+
+def test_many_returns():
+    @ray.remote(num_returns=MAX_RETURNS)
+    def f():
+        to_return = []
+        for _ in range(MAX_RETURNS):
+            obj = list(range(10000))
+            to_return.append(obj)
+
+        return tuple(to_return)
+
+    returned_refs = f.remote()
+    assert len(returned_refs) == MAX_RETURNS
+
+    for ref in returned_refs:
+        expected = list(range(10000))
+        obj = ray.get(ref)
+        assert obj == expected
+
+def test_ray_get_args():
+    def with_dese():
+        print("Putting test objects:")
+        refs = []
+        for _ in trange(MAX_RAY_GET_ARGS):
+            obj = list(range(10000))
+            refs.append(ray.put(obj))
+
+        print("Getting objects")
+        results = ray.get(refs)
+        assert len(results) == MAX_RAY_GET_ARGS
+
+        print("Asserting correctness")
+        for obj in tqdm(results):
+            expected = list(range(10000))
+            assert obj == expected
+
+    def with_zero_copy():
+        print("Putting test objects:")
+        refs = []
+        for _ in trange(MAX_RAY_GET_ARGS):
+            obj = np.arange(10000)
+            refs.append(ray.put(obj))
+
+        print("Getting objects")
+        results = ray.get(refs)
+        assert len(results) == MAX_RAY_GET_ARGS
+
+        print("Asserting correctness")
+        for obj in tqdm(results):
+            expected = np.arange(10000)
+            assert (obj == expected).all()
+
+    with_dese()
+    print("Done with dese")
+    with_zero_copy()
+    print("Done with zero copy")
+
+def test_many_queued_tasks():
+    sema = Semaphore.remote(0)
+    @ray.remote(num_cpus=1)
+    def block():
+        ray.get(sema.acquire.remote())
+
+    @ray.remote(num_cpus=1)
+    def f():
+        pass
+
+    num_cpus = int(ray.cluster_resources()["CPU"])
+    blocked_tasks = []
+    for _ in range(num_cpus):
+        blocked_tasks.append(block.remote())
+
+    print("Submitting many tasks")
+    pending_tasks = []
+    for _ in trange(MAX_QUEUED_TASKS):
+        pending_tasks.append(f.remote())
+
+    # Make sure all the tasks can actually run.
+    for _ in range(num_cpus):
+        sema.release.remote()
+
+    print("Unblocking tasks")
+    for ref in tqdm(pending_tasks):
+        assert ray.get(ref) is None
+
+def test_large_object():
+    # int is 4 bytes
+    num_elements = MAX_RAY_GET_SIZE // 4
+
+    print("Putting object")
+    ref = ray.put([2**31 for _ in range(num_elements)])
+    print("Getting object")
+    big_obj = ray.get(ref)
+
+    print("Verifying object")
+    for val in tqdm(big_obj):
+        assert val == 2**31
+
+
+ray.init(address="auto")
```

Comment:
Are we sharing the same cluster to test all?

Reply:
Yeah these tests are all very fast (<5 min to run all of them sequentially) so the cost of starting multiple instances would probably outweighs the cost of actually running the tests.

Reply:
Any possibility the previous test affects the later tests?

Reply:
They shouldn't (and if they do, I'd argue thats a bug)

Reply:
Don’t we need to add some check at least to ensure all metadata is cleaned?

Reply:
what do you mean by metadata?

Reply:
Ah I just meant the resources are correctly returned and things like that.

Reply:
oh i'm not sure if we should start including too many correctness checks that should be covered by unit tests. I mostly used assert statements on correctness to make sure we were measuring the right things (to prevent potentially strange results from things like lazy loading optimizations).

Reply:
But in that case, we don't know if the issue is due to the previous test affecting the next one or the test itself is broken right? I agree if it is not working, that's a bug, but don't we need a way to detect a bug at least (and make sure tests are not failed due to other factors)?

Reply:
ok, i can assert that the resources are properly reset, but i would think any bug should be caught/detected by unit tests not by this

Reply:
Yeah I definitely agree.